File: BOJ/1780/A.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<vector<int>> arr;

vector<pair<int, int>> dir{{0, 0}, {0, 1}, {0, 2}, {1, 0}, {1, 1}, {1, 2}, {2, 0}, {2, 1}, {2, 2}};

vector<int> divide(int size, int x, int y)
{
    if (size == 1)
    {
        int idx = arr[x][y] + 1;

        vector<int> r(3, 0);
        r[idx] = 1;
        return r;
    }
    else
    {
        vector<vector<int>> div;
        int s = int(size / 3);
        vector<int> ans(3, 0);
        for (int i = 0; i < dir.size(); i++)
        {
            vector<int> temp = divide(s, x + dir[i].first * s, y + dir[i].second * s);
            div.push_back(temp);
        }
        for (int i = 0; i < 9; i++)
        {
            for (int j = 0; j < 3; j++)
            {
                ans[j] += div[i][j];
            }
        }
        for (int i = 0; i < 3; i++)
        {
            if (ans[i] == (size * size) / (s * s))
            {
                ans[i] = 1;
            }
        }
        return ans;
    }
}

int main()
{
    freopen("input.txt", "rt", stdin);
    ios::sync_with_stdio(0);
    cin.tie(0);
    int N;
    cin >> N;
    for (int i = 0; i < N; i++)
    {
        vector<int> temp;
        for (int j = 0; j < N; j++)
        {
            int n;
            cin >> n;
            temp.push_back(n);
        }
        arr.push_back(temp);
    }
    vector<int> result = divide(N, 0, 0);
    cout << result[0] << "\n"
         << result[1] << "\n"
         << result[2];
}

File: BOJ/1780/A.py
import sys

sys.stdin = open("D:\Projects\AA\BOJ\\1780\input.txt")
input = sys.stdin.readline

N = int(input().rstrip())

arr = []

for i in range(N):
    temp = list(map(int, input().rstrip().split()))
    arr.append(temp)


def divide(x, y, size):
    if size == 1:
        num = arr[x][y]
        result = []
        if num == 1:
            result = [0, 0, 1, True]
        elif num == -1:
            result = [1, 0, 0, True]
        else:
            result = [0, 1, 0, True]
        return result
    else:
        result = [0, 0, 0, True]
        smaller_size = size//3
        vals = 0
        first = divide(x, y, smaller_size)
        for i in range(3):
            if first[i]:
                vals += 1
        if vals != 1:
            result[3] = False
        for i in range(3):
            for j in range(3):
                temp = divide(x+i*smaller_size, y+j *
                              smaller_size, smaller_size)
                if temp != first:
                    result[3] = False
                for k in range(3):
                    result[k] = result[k] + temp[k]
        if result[3]:
            result = first
        return result


ans = divide(0, 0, N)

for i in range(3):
    print(ans[i])
